### DISCUSSION TITLE
PoC Aufgabe: RegEx Monster

### DISCUSSION THREAD

**Discussion by AMatutat** [Programmiermethoden/Dungeon#83] [Announcements]
Diese Aufgabe soll mit dem, in https://github.com/Programmiermethoden/Dungeon/discussions/81 besprochenen, Pattern umgesetzt werden und als proof of concept dienen. 

## Referenzaufgabe ohne Dungeonbezug
**Bilde den Regulären Ausdruck**
Schreiben Sie einen regulären Ausdruck, der E-Mail Adressen nach dem bekannten Schema widerspiegelt. 
Beachten Sie dabei: 
- Eine E-Mail Adresse darf nicht mit einer Zahl oder einen Sonderzeichen beginne.
- Vor dem `@` dürfen beliebig viele Zahlen, Buchstaben und Sonderzeichen stehen
- Nach dem `@` dürfen maximal 5 kleine Buchstaben, minimum aber 3 folgen, danach folgt ein `.``
- Nach dem `.`dürfen maximal 3 kleine Buchstaben folgen, minimum aber 2

**Regex lesen:**
Schreiben Sie eine gültige Eingabe für folgenden Regulären Ausdruck `^((\\+1)?\\s?\\(\\d{3}\\)\\s?\\d{3}\\-\\d{4})?$")`

@cagix Du hast hier bestimmt irgendwo bessere Aufgaben, oder? 

## Spielablauf
 
Im Spiel laufen die bekannten Monster herum. Einige Monster sind aber besonders gefährlich, es sind nämlich `RegEx-Monster` (bibber). 
Kommt es zum Kampf mit so einem Monster, muss der Spieler eine RegeEx-Frage beantworten.
Dafür pausiert das Spiel, eine Textbox mit der Frage (z.B eine von Oben) und eine Eingabemöglichkeit erscheint.
Bei einer richtigen Eingabe wird das Monster besiegt, bei einer falschen Eingabe bekommt der Spieler Schaden und muss es erneut probieren. 
Besiegte `RegEx-Monster` lassen wertvollen Loot fallen, es lohnt sich also. 

## Bewertung

Die Bewertung kann Anhand der Fehlversuche berechnet werden. Auch wäre es möglich, verschiedene Lösungen anzubieten (diese müsste man dann irgendwo definieren) und diese untereinander zu Werten. 
Für die Eingabe von eigenen RegEx braucht es mehrere Text-Werte, um statische RegEx Eingaben zu erkennen.

**Comment by malt-r** [comment]
# DSL Eingabe

## Erster Entwuf

Annahme: Held*in wird mit einem String, der nach einem Pattern generiert wird, angegriffen und muss ein passendes Regex-Pattern in eine Textbox eingeben.

Wenn das Monster mit dem richtigen Pattern besiegt wird, lässt es Loot aus der Inventory-Komponente fallen.

```
fn drop_on_die(gameobject died) {
  // using this in eventHandler of gameobject, which has no inventory_comp would 
  // lead to runtime error -> could we check this semantically?
  // DSL will be using static typesystem, so this could be checked theoretically for each call of drop_on_die
  // because we know, which gameobject has which components associated with it in semantic analysis
  died.inventory_comp.drop_all();
}

let pattern = "[a-z][A-Z]+"

fn attack_with_regex(gameobject me, gameobject opponent) {
  let str = get_string_from_regex(pattern);
  let task_str = "Give me a regex for \"$pattern\" or die!";
  let answer_pattern = show_textbox(task_str); // parameter is task description

  if match_regex(answer_pattern, str) {
    show_text("Oh no, you killed me!");
    me.health_comp.set_health(0);
  } else {
    show_text("Ha, wrong!");
    oppponent.combat_comp.damage(); // pass own damage or use implicitly?
  }
}

// eigenes gameobject als Kombination aus Komponenten definieren, was dann instanziiert werden kann
gameobject regex_monster {
  combat_comp {
    health: 100,
    damage: 10,
    type: @MELEE, // binding an java enum-Variante, Syntax noch WIP
    on_attack: attack_with_regex, // override attack handler -> need to match signature
    on_die: drop_on_die           // drop loot from inventory_comp, would be really convenient, 
                                  // if this could be an anonymous function..
  },
  inventory_comp {
    contents: [<items>] // array definition
  },
  xp_container_comp {
    amount: 10
  },
  logging_comp
}

room quest_room {
  monsters: [regex_monster],
  // optional könnte hier noch ein flag gesetzt werden, was die automatische 
  // generierung von Füller-Content ein- oder ausschaltet (mehr monster)
  size: @BIG
}

quest_config qc {
  quest_points: 10,
  quest_desc: "Finde und besiege das Regex-Monster!"
  // falls kein level_graph vorgegeben wird, sollte einer zufällig erzeugt werden;
  // es sollte allerdings trotzdem die Möglichkeit gegeben sein, selber Räume 
  // zu definieren und im Dungeon platzieren zu lassen und dann weitere Räume
  // automatisch "drumherum" generieren zu lassen
  rooms: [quest_room],
  generate_rooms: true
}
```

## Offene Punkte

- Syntax für Übergabe von Java Enumvariante: könnte einerseits speziell gekennzeichnet sein (was es visuell für DSL-Nutzende besser kennzeichnen würde), andererseits könnte der übergebene Name (hier "MELEE") auch im Kontext der Komponente resolved werden, was die Syntax vereinfachen würde
